runstatus.go

```diff
@@ -0,0 +1,357 @@
+	urlSplited := strings.Split(url, "/")
+	id, err := strconv.ParseInt(urlSplited[len(urlSplited)-1], 10, 32)
+	if err != nil {
+		return 0, err
+	}
+
+	return int(id), nil
+}
+
+// DeleteRunstatusService delete runstatus service
+func (client *Client) DeleteRunstatusService(ctx context.Context, page string, id int) error {
+	_, err := client.runstatusRequest(ctx, fmt.Sprintf("/pages/%s/services/%d", page, id), nil, "DELETE")
+	return err
+}
+
+// CreateRunstatusService create runstatus service
+func (client *Client) CreateRunstatusService(ctx context.Context, page string, service RunstatusService) error {
+	_, err := client.runstatusRequest(ctx, "/pages/"+page+"/services", service, "POST")
+	return err
+}
+
+// ListRunstatusService list runstatus service
+func (client *Client) ListRunstatusService(ctx context.Context, page string) ([]RunstatusService, error) {
+	resp, err := client.runstatusRequest(ctx, "/pages/"+page+"/services", nil, "GET")
+	if err != nil {
+		return nil, err
+	}
+
+	var p *RunstatusServiceList
+	if err := json.Unmarshal(resp, &p); err != nil {
+		return nil, err
+	}
+
+	return p.Services, nil
+}
+
+// CreateRunstatusEvent create runstatus incident event
+func (client *Client) CreateRunstatusEvent(ctx context.Context, page string, incidentID int, event RunstatusEvent) error {
+	_, err := client.runstatusRequest(ctx, fmt.Sprintf("/pages/%s/incidents/%d/events", page, incidentID), event, "POST")
+	return err
+}
+
+// ListRunstatusMaintenance list runstatus Maintenance
+func (client *Client) ListRunstatusMaintenance(ctx context.Context, page string) ([]RunstatusMaintenance, error) {
+	resp, err := client.runstatusRequest(ctx, "/pages/"+page+"/maintenances", nil, "GET")
+	if err != nil {
+		return nil, err
+	}
+
+	var p *RunstatusMaintenanceList
+	if err := json.Unmarshal(resp, &p); err != nil {
+		return nil, err
+	}
+
+	return p.Maintenances, nil
+}
+
+// CreateRunstatusMaintenance create runstatus Maintenance
+func (client *Client) CreateRunstatusMaintenance(ctx context.Context, page string, maintenance RunstatusMaintenance) error {
+	_, err := client.runstatusRequest(ctx, "/pages/"+page+"/maintenances", maintenance, "POST")
+	return err
+}
+
+// DeleteRunstatusMaintenance delete runstatus Maintenance
+func (client *Client) DeleteRunstatusMaintenance(ctx context.Context, page string, id int) error {
+	_, err := client.runstatusRequest(ctx, fmt.Sprintf("/pages/%s/maintenances/%d", page, id), nil, "DELETE")
+	return err
+}
+
+// ID give the maintenance ID
+func (service *RunstatusMaintenance) ID() (int, error) {
+	url := strings.TrimRight(service.URL, "/")
+	urlSplited := strings.Split(url, "/")
+	id, err := strconv.ParseInt(urlSplited[len(urlSplited)-1], 10, 64)
+	if err != nil {
+		return 0, err
+	}
+
+	return int(id), nil
+}
+
+// UpdateRunstatusMaintenance update runstatus Maintenance
+func (client *Client) UpdateRunstatusMaintenance(ctx context.Context, page string, id int, event RunstatusEvent) error {
```

Comment:
I wonder if it wouldn't be nicer as a method on the `RunstatusMaintenance` type:

```go
func (m *RunstatusMaintenance) Update(ctx context.Context, event RunstatusEvent) error
```

The `page` parameter could be a `RunstatusMaintenance` struct field?

Reply:
yes it would be nicer But where I put the egoscale `client`, the `page` name to know which page to modify and where I can put the `id` of the maintenance to know which event I have to modify, knowing that id is not naturally in maintenance struct.

Reply:
`RunstatusMaintenance` cannot let me do that. Or maybe a PR in runstat.us can fix it

Reply:
I kind of agree with @falzm here. We probably have a HATE-OAS problem in runstatus.

Reply:
I'm agree too but as you said the problem come from runstatus

Reply:
If the Runstat.us data model allows it, I suggest a more fluid API style such as:

```go
type RunstatusPage struct {
        // all current fields, plus:
        client *Client
}

func (client *Client) GetRunstatusPage(ctx context.Context, pageName string) (*RunstatusPage, error) {}

func (page *RunstatusPage) CreateIncident(ctx context.Context, incident RunstatusIncident) error {}

func (page *RunstatusPage) CreateMaintenance(ctx context.Context, maintenance RunstatusMaintenance) error {}

func (m *RunstatusMaintenance) Update(ctx context.Context, event RunstatusEvent) error {}
```

WDTY?

Reply:
👍 LGTM

Reply:
So, not going in that direction after all?

Reply:
Embedding the client/page has been done in a more superficial way by hardcoding the `/pages` URL and otherwise using `PageURL`.

```go
client.CreateIncident(ctx, RunstatusIncident{PageURL: page.URL, ...})
// vs
page.CreateIncident(ctx, RunstatusIncident{...})
```

It feels more consistent with the rest of the library.